Proj/xld pull request

stephlazar opened the pull request:
There are conflicts between Proj/XLD and qudev_master, but I still create the pull request so we can review changes.

Changes proposed in this pull request:
- latest developments in analysis_v3
- renames UHFQC_integration_average_classifier_det to UHFQC_classifier_detector
- renames find_frequency (for qb spec) to find_qubit_frequency
- makes MC.run first save metadata then run experiment
- makes classifier detector work with multiple UHFLIs
- adds measure_ramsey_with_flux_pulse
- makes push-away flux pulse unipolar even when using NZ pulse.

@antsr 
@nathlacroix

Commit:
Extracts measured data based in sweep_points, not hard_sweep_points and soft_sweep_points.
Changes data_keys_in, data_keys_out to keys_in, keys_out.
Adds get_cp_sp_spmap_measobjn to data_processing.py.
Fixes bugs in rotate_1d_array.
Adds option to call fitting_and_plotting.py/plot in prepare_1d_plot.
Adds keys_in parameter to run_fitting, plot, save_figures.
Adds fit_one_dict to fitting_and_plotting.py.

antsr committed:
Improved surface code measurement, fixed prefetch for uhf

Commit:
Fixes bug in fitting_and_plotting.py where save_figures is called with the wrong keys_in.
Other minor changes.

Commit:
Fixes bug in MultiQubit_TimeDomain_Analysis related to cal_points, and adds normalization of
   data if one RO channe and no cal points.
Updates the T1 measurements to new wfm generation.
Adds measure_ramsey_with_flux_pulse to multi_qubit_module.py.
Adds Ramsey_with_flux_pulse_meas_seq.
Updates ro_dynamic_phase_seq to new wfm generation framework.
Renames T1_seq to t1_active_reset and updates it to new wfm generation.

Commit:
Adds functionality to take all keys in measured_data if keys_in == None in the processing nodes.
Adds option to pass in exp_metadata in options_dict.

Commit:
Splits fitting and plotting modules into fitting.py and plotting.py.
Adds helper_functions.py.
Adds prepare_cos_fit_dict to fitting.py.

Commit:
Renames base_analysis.py to pipe_analysis.py.
Fixes bug when exact_label_match == True in analysis/analysis_toolbox.py get_timestamps_in_range.
Adds prepare_1d_plot_dicts, prepare_raw_data_plot_dicts, prepare_cal_states_plot_dicts to plotting.py,
   and adds calls to these functions in RabiAnalysis and SingleQubitRBAnalysis.
Removes prepare_1d_plot from data_processing.py.
Modifies helper_functions.py get_params to check the data_dict before checking experimental metadata.

Commit:
Renames BaseDataAnalysis class to PipeDataAnalysis.

Commit:
Cleans up process_data in PipeDataAnalysis.
Modifies filter_data to accept both strings and functions in data_filter_func.
Adds raw data plotting both before and after filtering in RabiAnalysis and SingleQubitRBAnalysis.

Commit:
Modifies the run method of MeasurementControl to save metadata at the beginning of measurement.

Commit:
Merge branch 'devel/Cphi_gate' into devel/analysis_v3

Commit:
Adds thresholding option to UHFQC_integration_average_classifier_det.

Commit:
Adds processing_pipeline.py.
Renames pipe_analysis to pipeline_analysis.py.
Renames num_avg_bins to num_bins in average and get_std_deviation nodes.
Adds do_simple_fit parameter to SingleQubitRBAnalysis.
Modifies prepare_raw_data_plot_dicts to accept keys_in as argument.

Commit:
Adds get_meas_obj_value_names_map to multi_qubit_module.py
Adapts measure_two_qubit_randomized_benchmarking and measure_randomized_benchmarking
   to using analysis_v3 with all the current features.
Improves analysis_v3 to work even if a msmt_obj_name is not in cal_points.
Adds to feature to add a general node to ProcessingPipeline (not found among methods of the class).
Removes automatic finding of std keys i add_SingleQubitRBAnalysis_node.
Changes value_names in UHFQC_correlation_detector to match those in UHFQC_integrated_average_detector.
Adds result_logging_mode as input paramters to UHFQC_correlation_detector and sets it to UHF.

Commit:
Fixes bug in measure_two_qubit_randomized_benchmarking in adding the correlation channel to
   meas_obj_value_names_map.

Commit:
Renames find_frequency to find_qubit_frequency.
Renames the classifier detector to UHFQC_classifier_detector.

Commit:
Adds measure_resonator_amplitude_sweep (doesn't work).
Adds res_amp_sweep_seqs.

antsr committed:
Merge branch 'devel/analysis_v3' into Proj/XLD

# Conflicts:
#	pycqed/analysis_v2/timedomain_analysis.py
#	pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py
#	pycqed/measurement/multi_qubit_module.py
#	pycqed/measurement/pulse_sequences/multi_qubit_tek_seq_elts.py

Commit:
Fixes bug in find_ssro_fidelity for qubit: nr of sweep points was being multiplied by nr of shots.
Simplifies the code for thresholding data in the UHFQC_classifier_detector.
Fixes bug in MeasurementControl create_experimentaldata_dataset, where it was not checking whether
   'Experimental Data' group exists in HDF5 file.
Fixes bug in get_multiplexed_readout_detector_functions: the acq_classifier_params and acq_state_prob_mtxs
   were not added per UHF.
Fixes bug in measure_cphase where the channel_map was created from the value names of the incorrect detector function.

Commit:
Adds metadata extraction to MeasurementAnalysis.
Fixes FluxPulse_Scope_Analysis to use physical sweep points.
Adds self.data_with_reset flag to MultiQubit_TimeDomain_Analysis, and checks it in prepare_raw_data_plots.
Adds preparation params to experimental metadata in measure_qscale.
Adds correct label suffix to measure_cphase.
Makes push-away flux pulse a unipolar pulse even for the NZBufferedCZPulse.

Commit:
Removes print statements, replaces some with log.info and log.warning.

Commit:
Fixes bug in get_naming_and_values and get_naming_and_values_2D where we got a KeyError
   if 'Experimental Metadata' group did not exist in Experimental Data group.

Commit:
Merge branch 'Proj/XLD' into Proj/XLD_pull_request

nathlacroix commented:
FYI : 
[https://stackoverflow.com/questions/43547323/is-a-feature-branch-merged-into-the-master-branch-or-the-opposite](https://stackoverflow.com/questions/43547323/is-a-feature-branch-merged-into-the-master-branch-or-the-opposite) 
So we should merge master into XLD, resolve conflicts and then merge back into master :) (not the opposite)

Commit:
Merge branch 'qudev_master' into Proj/XLD_pull_request

# Conflicts:
#	pycqed/analysis_v2/readout_analysis.py
#	pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py
#	pycqed/measurement/awg_sweep_functions.py
#	pycqed/measurement/detector_functions.py
#	pycqed/measurement/multi_qubit_module.py
#	pycqed/measurement/pulse_sequences/multi_qubit_tek_seq_elts.py
#	pycqed/measurement/pulse_sequences/single_qubit_tek_seq_elts.py

antsr commented on pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py:
```diff
@@ -888,9 +946,10 @@ def measure_qscale(self, qscales=None, analyze=True, upload=True, label=None,
                              'preparation_params': prep_params,
              'sweep_name': 'Qscale factor',
              'sweep_unit': '',
+             'preparation_params': prep_params,
```
Duplicate parameter

nathlacroix replied:
duplicate line of 946?

antsr commented on pycqed/measurement/detector_functions.py:
```diff
@@ -837,20 +839,21 @@ class UHFQC_correlation_detector(UHFQC_integrated_average_detector):
             channels=channels,
             seg_per_point=seg_per_point, single_int_avg=single_int_avg,
-            result_logging_mode='raw',  # FIXME -> do the proper thing (MAR)
+            result_logging_mode=result_logging_mode,  # FIXME -> do the proper thing (MAR)
```
I wonder if this fixme is now fixed.

stephlazar replied:
It is now :) Bu this detector function has not been used in some time so it probably doesn't even work anymore.

nathlacroix commented on pycqed/analysis/measurement_analysis.py:
```diff
@@ -681,6 +682,11 @@ def get_naming_and_values(self, new_sweep_points=None):
             self.sweep_points_from_file = self.sweep_points
             self.sweep_points = new_sweep_points
 
+        try:
+            self.exp_metadata = h5d.read_dict_from_hdf5(
+                {}, self.data_file['Experimental Data']['Experimental Metadata'])
+        except KeyError:
+            self.exp_metadata = {}
```
Will this fix the issue discussed in my pull request of receiving the `[]` instead of `{}` for the experimental metadata?

stephlazar replied:
No, this is in analysis_v1, your issue was in v2.

nathlacroix commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -334,14 +336,16 @@ def process_data(self):
         if self.cal_states_rotations is not None:
             self.cal_states_analysis()
         else:
+            # this assumes data obetained with classifier detector!
+            # ie pg, pe, pf are expected to be in the value_names
```
Could we check for this and add an error if it is not the case ? or is there no way of doing that ? Because I think I ran in this issue more than once and it is not that easy to trace it back (you get a key error somewhere downstream).

stephlazar replied:
I don't know if I would spend time on improving analysis_v2; this should encourage people to switch to v3 and help with development!

antsr commented on pycqed/measurement/multi_qubit_module.py:
```diff
@@ -2875,62 +3003,79 @@ def measure_pygsti(qubits, f_LO, pygsti_gateset=None,
+    if exp_metadata is None:
+        exp_metadata = {}
+
+    hard_sweep_params = {
```
@Vintrix Do you understand what is going on here, why are the hard sweep parameters now in degrees, while before they were in rad. Also they are hardcoded here. Also now they are repeated and tiled while before only tiled. This function in its current state I would not want in the qudev_master. We could also just delete it for now and keep it only in Proj/XLD, if its hard to fix.

stephlazar replied:
Yes, let's do that for now.

nathlacroix commented on pycqed/analysis_v3/data_processing.py:
```diff
@@ -1,232 +1,104 @@
 import logging
 log = logging.getLogger()
 log.addHandler(logging.StreamHandler())
```
Loggers should have a name, ie `__name__`. Creating several of  those loggers without name (ie. root loggers) might be responsible for our log duplicates (but I m not sure about that).

antsr commented on pycqed/measurement/pulse_sequences/multi_qubit_tek_seq_elts.py:
```diff
@@ -1773,11 +1802,11 @@ def pygsti_seq(qb_names, pygsti_listOfExperiments, operation_dict,
 
-def ro_dynamic_phase_seq(qbp_name, qbr_names,
-                         phases, operation_dict,
+def ro_dynamic_phase_seq(hard_sweep_dict, qbp_name, qbr_names,
```
The updates to this function need some cleaning up. Or we just reject the changes for now.

stephlazar replied:
Completely removed from master. Should be put into master when it is cleaned up.

antsr commented on pycqed/measurement/pulse_sequences/single_qubit_tek_seq_elts.py:
```diff
@@ -833,6 +834,35 @@ def over_under_rotation_seq(qb_name, nr_pi_pulses_array, operation_dict,
     return
 
 
+def res_amp_sweep_seqs(qb_name, hard_sweep_dict, soft_sweep_dict,
```
This function is really cryptic to me.

antsr replied:
Is it some general readout parameter 2d sweep sequence?

antsr replied:
@nathlacroix ?

nathlacroix replied:
same for me, A doc string might be helpful, as I have commented in my review

stephlazar replied:
The measurement corresponding to this sequence doesn't work anyway. I should have removed it. I will do so now, along with the measure function.

nathlacroix commented on pycqed/analysis_v3/data_processing.py:
```diff
@@ -577,318 +440,136 @@ def rotate_iq(data_dict, data_keys_in, data_keys_out, **params):
     return data_dict
 
 
-## Helper functions ##
```
I see there is now a module for helper functions. I think that is indeed better than having them inside another file. Yet it might be useful to have meaningful names for the relevant helper submodules, otherwise it might end like the analysis toolbox where there is plenty of things but you don't know what exactly or what the structure is (or there simply is no structure hahaha). This is not a comment specially relevant for the pull request but as I'm reading the code I'm just wondering what structure we want for this helper module.

stephlazar replied:
I agree. Let's keep this in mind as we develop. For now there are only a few functions there ;)

antsr commented on pycqed/measurement/waveform_control/pulse_library.py:
```diff
@@ -456,32 +456,51 @@ def chan_wf(self, chan, tvals):
+                    (tvals - tend) * scaling)) * amp
+            t_rel = tvals - tvals[0]
+            wave *= np.cos(
+                2 * np.pi * (self.frequency * t_rel + self.phase / 360.))
```
The pushaway pulse has been changed to a modulated square pulse. This seems wrong. @Vintrix

Vintrix replied:
I agree :) I copied the wrong thing.

nathlacroix commented on pycqed/analysis_v3/data_processing.py:
```diff
@@ -577,318 +440,136 @@ def rotate_iq(data_dict, data_keys_in, data_keys_out, **params):
-        data_dict['plot_dicts'].update(plot_dicts)
-    else:
-        data_dict['plot_dicts'] = plot_dicts
-
```
Other general comment on analysis V3: Don't we want a base class for Node ? or would this be meaningless?

stephlazar replied:
Since the nodes can be either functions of classes, I don't know if this makes sense. Then you would be enforcing that they are all classes, but most should be functions in the original concept in order to keep things simple. The only things the nodes should have in common are a few input arguments.

nathlacroix commented on pycqed/analysis_v3/data_processing.py:
```diff
@@ -577,318 +440,136 @@ def rotate_iq(data_dict, data_keys_in, data_keys_out, **params):
-    if len(cp_list) != len(data_to_proc_dict):
-        raise ValueError('cal_points_list and data_keys_in do not have '
+
+    cp_list = help_func_mod.get_param('cal_points_list', data_dict, **params)
```
I'm not sure I understand why you would need a list of calibration point objects. The cal point object handles multi qubits. I don't think we should create more than 1 per experiment, or am I misunderstanding something ?

But if we keep it for now, does this mean that we have to recreate all the calpoint objects in the notebook if we want to do the analysis in the notebook? of course it's fine for now, but is that the end goal we want ?

stephlazar replied:
Good point, cp should not be a list. I will address this when I do a pull request from analysis_v3. 

You would not need to provide the cp in the notebook if they exist in the metadata. It will take them from there and convert them to a list.

nathlacroix commented on pycqed/analysis_v3/fitting.py:
```diff
@@ -0,0 +1,123 @@
+import logging
+log = logging.getLogger()
+log.addHandler(logging.StreamHandler())
```
logger needs name

nathlacroix commented on pycqed/analysis_v3/helper_functions.py:
```diff
@@ -0,0 +1,282 @@
+import logging
+log = logging.getLogger()
+log.addHandler(logging.StreamHandler())
```
logger needs name

nathlacroix commented on pycqed/analysis_v3/data_processing.py:
```diff
@@ -1,232 +1,104 @@
 import logging
```
General comment: Good job for the documentation ! It's really nice to have clear and documented nodes :) :+1:

nathlacroix replied:
I think I might be useful to make this a module with different node grouped in files, I already see this file becoming as large as the QudevTransmon when I arrived ;)

stephlazar replied:
Yep, that is the plan. I already split is into fitting, plotting, and helper functions modules. Originally I think most of those nodes were all in the data_processing.

nathlacroix commented on pycqed/analysis_v3/pipeline_analysis.py:
```diff
@@ -11,14 +11,17 @@
 import copy
 import logging
 log = logging.getLogger()
 log.addHandler(logging.StreamHandler())
```
logger needs name

nathlacroix commented on pycqed/analysis_v3/fitting.py:
```diff
@@ -0,0 +1,123 @@
+import logging
```
Good to have this in a separate file :)

stephlazar replied:
Have what in a separate file?

nathlacroix commented on pycqed/analysis_v3/plotting.py:
```diff
@@ -2,22 +2,18 @@
 log = logging.getLogger()
 log.addHandler(logging.StreamHandler())
```
logger needs name

nathlacroix commented on pycqed/analysis_v3/processing_pipeline.py:
```diff
@@ -0,0 +1,147 @@
+import re
+import logging
+log = logging.getLogger(__name__)
```
:+1:

nathlacroix commented on pycqed/analysis_v3/processing_pipeline.py:
```diff
@@ -0,0 +1,147 @@
+        #             # print(f'Taking keys_in as the keys_out of node '
+        #             #       f'{d["node_type"]} for the key suffix {keyi}')
+        #     if len(keys_out_in_pipe) > 0:
+        #         keys_in = [kk for k in keys_out_in_pipe for kk in k]
```
needed ?

stephlazar replied:
Not really

nathlacroix commented on pycqed/analysis_v3/saving.py:
```diff
@@ -164,7 +164,7 @@ def save_fit_results(data_dict, fit_res_dict, **params):
 
 def save_figures(data_dict, figs, **params):
 
-    key_list = params.get('key_list', 'auto')
```
logger needs name (could not place comment at top of file because it's not in the diff)

nathlacroix commented on pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py:
```diff
@@ -1284,28 +1327,21 @@ def measure_randomized_benchmarking(
         sp.add_sweep_dimension()
         sp.add_sweep_parameter('cliffords', cliffords, '',
                                'Number of applied Cliffords, $m$')
+        # create analysis pipeline object
```
I like this notation ! Would make sense to also have `pp.add_nodes(node1, node2, ...)`

stephlazar replied:
Yep, this is already implemented :)

nathlacroix commented on pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py:
```diff
@@ -2432,7 +2463,7 @@ def find_T1(self, times, n_cal_points_per_state=2, cal_states='auto',
+        # Extract T1 and T1_stddev from ma.T1_Analysis
         if analyze:
             T1_ana = tda.T1Analysis(qb_names=[self.name])
             if update:
```
line 2476: should have one more tab, otherwise if analyze is false, T1 is undefined.

nathlacroix commented on pycqed/measurement/measurement_control.py:
```diff
@@ -160,6 +160,8 @@ def run(self, name: str=None, exp_metadata: dict=None,
 
         with h5d.Data(name=self.get_measurement_name(),
                       datadir=self.datadir()) as self.data_object:
+            if exp_metadata is not None:
+                self.save_exp_metadata(exp_metadata, self.data_object)
```
:+1:

nathlacroix commented on pycqed/measurement/multi_qubit_module.py:
```diff
@@ -162,14 +165,14 @@ def get_multiplexed_readout_detector_functions(qubits, nr_averages=2**10,
                                                used_channels=None,
                                                correlations=None,
                                                add_channels=None,
-                                               correlated=False,
```
This is then True by default ? where can we decide on that?

nathlacroix commented on pycqed/measurement/multi_qubit_module.py:
```diff
@@ -23,15 +23,18 @@
 import pycqed.measurement.detector_functions as det
```
Logger needs name, I think we should refactor that into all files!

stephlazar replied:
I fixed it now by hand

nathlacroix commented on pycqed/measurement/pulse_sequences/fluxing_sequences.py:
```diff
@@ -205,57 +205,73 @@ def dynamic_phase_seq(qb_name, hard_sweep_dict, operation_dict,
+                            cal_points=None,
+                            upload=False, prep_params=dict()):
+    '''
+    Performs a Ramsey with interleaved Flux pulse
```
thanks for the nice doc :)

nathlacroix commented on pycqed/measurement/pulse_sequences/multi_qubit_tek_seq_elts.py:
```diff
@@ -74,8 +74,20 @@ def n_qubit_off_on(pulse_pars_list, RO_pars_list, return_seq=False,
         seg_list.append(seg)
         seq.add(seg)
+
+    uhfs_used = dict()
+    for RO_pars in RO_pars_list:
+        uhf = RO_pars['I_channel'][:4]
+        if uhf not in uhfs_used.keys():
+            uhfs_used[uhf] = []
+        uhfs_used[uhf].append(1)
+
+    repeat_dict = {}
+    for uhf in uhfs_used:
+        repeat_dict[uhf] = ((1.0 + int(preselection))*len(pulse_combinations),1)
+
     if upload:
-        ps.Pulsar.get_instance().program_awgs(seq)
+        ps.Pulsar.get_instance().program_awgs(seq, repeat_dict=repeat_dict)
     if return_seq:
         return seq, seg_list
```
needs to be updated to new way of using repeat dict :)

stephlazar replied:
Function was removed.

nathlacroix commented on pycqed/measurement/pulse_sequences/multi_qubit_tek_seq_elts.py:
```diff
@@ -1017,6 +1040,11 @@ def multi_parity_multi_round_seq(ancilla_qubit_names,
+                            # all_pulses[-1]['pulse_delay'] = round*parity_seperation \
+                            #                                 + (1+rounds)*\
+                            #                                 m*parity_seperation +\
+                            #                                 (16/2.4)*150e-9
```
needed ?

stephlazar replied:
I don't know...that function is super messy..
@antsr

antsr replied:
Not sure about the commented code here. Christian wrote it. I would either leave the code as it is or delete both the measure and sequence functions from master.

stephlazar replied:
Yep, I deleted both. We will deal with it when Christian gets back.

nathlacroix commented on pycqed/measurement/pulse_sequences/multi_qubit_tek_seq_elts.py:
```diff
@@ -1042,12 +1070,15 @@ def multi_parity_multi_round_seq(ancilla_qubit_names,
                                  f'_loop_{parity_loops-1}_tomo_{t}'
-            pulse['pulse_delay'] = 300e-9
-            pulse['ref_point'] = 'end'
+            pulse['pulse_delay'] = (16/2.4)*168e-9 - \
```
would be even better if those numbers had names to that we know what they do :)

stephlazar replied:
I agree, want to add them? :))

stephlazar replied:
We now removed this function.

nathlacroix commented on pycqed/measurement/pulse_sequences/multi_qubit_tek_seq_elts.py:
```diff
@@ -1082,7 +1113,8 @@ def multi_parity_multi_round_seq(ancilla_qubit_names,
+                 (parity_loops, ROs),
+                 1
                  )
     log.debug(repeat_dict)
```
I guess the line splitting is not needed, but it doesn't bother either

nathlacroix commented on pycqed/measurement/pulse_sequences/single_qubit_tek_seq_elts.py:
```diff
@@ -833,6 +834,35 @@ def over_under_rotation_seq(qb_name, nr_pi_pulses_array, operation_dict,
 
+def res_amp_sweep_seqs(qb_name, hard_sweep_dict, soft_sweep_dict,
+                       uhf_name, operation_dict, upload=True):
+    seq_name = 'Res_amp_sweep_sequence'
```
docstring ? :)

stephlazar replied:
The function was removed.

nathlacroix commented on pycqed/measurement/waveform_control/pulse_library.py:
```diff
@@ -456,32 +456,51 @@ def chan_wf(self, chan, tvals):
+            # buffer_start -= self.extra_buffer_aux_pulse
+            # buffer_end -= self.extra_buffer_aux_pulse
+            # pulse_length += 2*self.extra_buffer_aux_pulse
+            # l1 = self.alpha*pulse_length/(self.alpha + 1)
```
needed ?

stephlazar replied:
This was fixed.

nathlacroix reviewed (changes requested):
A lot of cool new features ! looking forward to having a level up master :1st_place_medal: 
Not all comments are relevant to the pull request (several are more conceptual) but I think it makes sense to have them written down to start discussions about it :)

Commit:
Addresses pull request requested changes.

- removes the resonator amplitude sweep and ro dynamic phase measurements + sequence functions.
- properly implements the result_logging_mode in UHFQC_correlation_detector.
- reverts push-away flux pulse to be nz for the NZBufferedCZPulse.

Commit:
Addresses Natha's requested changes from the pull request.
- adds __name__ to Logger in analysis_v3 files and multi_qubit_module.py.
- removes measure_multi_parity_multi_round measure and sequence functions.
